### PR TITLE
Fix enclosure merging when using Construction:AirBoundary

### LINKED ISSUE #9218: EnergyPlus Crashes Without Error Log When Several Zones Are Connected by Construction:AirBoundary
Issue overview
--------------
I and several users of [Ladybug Tools](https://www.ladybug.tools/) have been making heavy use of the `Construction:AirBoundary` that was introduced to EnergyPlus a few versions ago, which has been a godsend and has allowed us to model many situations that we could not model previously. However, we have discovered that, in large models with big contiguous spaces connected by `Construction:AirBoundary`, EnergyPlus will crash during the "Initializing Simulation" phase. The .err file will be completely empty in this case and there's no record of what exactly when wrong. When the `Construction:AirBoundary` is replaced by just a normal opaque `Construction`, the simulation is successful.

### Details
We are using [the last official release of EnergyPlus 9.6](https://github.com/NREL/EnergyPlus/releases/tag/v9.6.0) and here you can find an IDF exported for EnergyPlus 9.6 that can recreate the crash:
https://drive.google.com/file/d/1NJWAkZzULieEp1RXAyY80dI3AflTX2xP/view?usp=sharing

I tried to create as minimal of a sample file as I could and I apologize that there are still 24 Zones and 216 Surfaces in the model. However, in our experience, we have only been able to recreate this error in cases when there are fairly large numbers of zones and `Construction:AirBoundary` surfaces in the model. To illustrate that the error is, in fact, the result of the `Construction:AirBoundary`, here is the exact same model using a normal opaque `Construction`, which can simulate successfully:
https://drive.google.com/file/d/1yNhpB_AhWyRE2i5KKjDEsM4D0FAwBpsK/view?usp=sharing

We have been able to reproduce the error on both Windows and Linux and across a wide range of system specifications. So this bug does not appear to be platform-specific.

In my opinion as an energy modeler, this type of use of `Construction:AirBoundary` is pretty excessive and I think most users of Ladybug Tools would find it acceptable if this excessive use of `Construction:AirBoundary` ultimately remains non-simulate-able. We can always join many of these Zones together into one and accept a higher degree of air mixing across the model. However, the fact that EnergyPlus crashes without any error log has been challenging for us since it forces us to question what exactly went wrong and whether the `Construction:AirBoundary` is, in fact to blame. So, at a minimum, we would really benefit from some error reporting that made it clear to us that the excessive use of `Construction:AirBoundary` is to blame.

### Checklist

 - [x] Defect file added (see above)
 - [ ] Ticket added to Pivotal for defect
 - [ ] Pull request created

### LINKED ISSUE #9233: Enclosure area and zone component loads incorrect when air boundary surface has the same space on both sides
Issue overview
--------------
If a Construction:AirBoundary surface has the same Space on both sides, the enclosure space list will include that space twice, and the floor area and other attributes of the enclosure will double-count that Space. This impacts the zone component loads report, because the enclosure floor area is used to scale the radiant pulse for delayed load components.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 9.6.0

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here) - see discussion of diffs [here](https://github.com/NREL/EnergyPlus/pull/9237#issuecomment-1016651362)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #9218
 - Fixes #9233 
 - Fixes radiant and solar enclosure logic to properly merge enclosures when an air boundary surface connects 2 enclosures for certain cases.
 - Fixes double-counting of space floor area and other properties when an air boundary surface has the same Space on both sides.

Initial commit is 2 unit tests which should fail with current code.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions - see [comments below](https://github.com/NREL/EnergyPlus/pull/9237#issuecomment-1016651362)
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] ~If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~
 - [ ] ~If structural output changes, add to output rules file and add OutputChange label~
 - [ ] ~If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mjwitte** [comment]
The diffs in 5ZoneAirCooledWithSpaces were unexpected. There are diffs in the component load decay curves, e.g.:
```
-Radiant to Convective Decay Curves for Cooling,ZONE 1,FRONT-1, 0.257, 0.158, 0.108, 0.076, 0.057, 0.044, 0.036, 0.030, 0.026, 0.023, 0.020, 0.018, 0.017, 0.015, 0.014, 0.013, 0.012, 0.011, 0.010, 0.010, 0.009, 0.009, 0.008, 0.007, 0.007, 0.007, 0.006, 0.006, 0.005, 0.005, 0.005, 0.005, 0.004, 0.004, 0.004, 0.004
-Radiant to Convective Decay Curves for Cooling,ZONE 1,WF-1, 0.241, 0.122, 0.048, 0.037, 0.030, 0.025, 0.021, 0.019, 0.016, 0.015, 0.013, 0.012, 0.011, 0.010, 0.010, 0.009, 0.008, 0.008, 0.007, 0.007, 0.006, 0.006, 0.005, 0.005, 0.005, 0.005, 0.004, 0.004, 0.004, 0.004, 0.003, 0.003, 0.003, 0.003, 0.003, 0.002
+Radiant to Convective Decay Curves for Cooling,ZONE 1,FRONT-1, 0.257, 0.158, 0.108, 0.076, 0.057, 0.044, 0.036, 0.030, 0.026, 0.023, 0.020, 0.018, 0.017, 0.015, 0.014, 0.013, 0.012, 0.011, 0.010, 0.010, 0.009, 0.008, 0.008, 0.007, 0.007, 0.007, 0.006, 0.006, 0.005, 0.005, 0.005, 0.005, 0.004, 0.004, 0.004, 0.004
+Radiant to Convective Decay Curves for Cooling,ZONE 1,WF-1, 0.241, 0.122, 0.048, 0.037, 0.030, 0.025, 0.021, 0.019, 0.016, 0.015, 0.013, 0.012, 0.011, 0.010, 0.009, 0.009, 0.008, 0.008, 0.007, 0.007, 0.006, 0.006, 0.005, 0.005, 0.005, 0.004, 0.004, 0.004, 0.004, 0.004, 0.003, 0.003, 0.003, 0.003, 0.003, 0.002
```
And big diffs in the table outputs for zone component loads. Why?

It turns out that `dataViewFactor->EnclRadInfo.FloorArea` was wrong for some enclosures. Debugger screenshots from before and after this fix show the floor area for Radiant Enclosure 1 (Solar Enclosure 1) was 126.36 before, and is 89.82 now:
image  image

29.88+59.94=89.82 is the correct area based on the Space Summary Report (which did not change before/after this fix).
image

So, why does the enclosure floor area impact the zone component loads report and nothing else? It appears that the zone component loads report is the only place that uses enclosure floor area. It's used to scale the pulse in `InitIntSolarDistribution` and `InitInternalHeatGains`, e.g.:
```
                    // for the loads component report during the special sizing run increase the radiant portion
                    // a small amount to create a "pulse" of heat that is used for the
                    state.dataHeatBalSurfMgr->adjQL =
                        state.dataHeatBalSurfMgr->curQL + state.dataViewFactor->EnclRadInfo(radEnclosureNum).FloorArea * pulseMultipler;
```
So, that's why.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 1]
```diff
@@ -14819,20 +14819,23 @@ namespace SurfaceGeometry {
                             thisEnclosure.spaceNums.push_back(surf.spaceNum);
                             thisEnclosure.FloorArea += state.dataHeatBal->space(surf.spaceNum).floorArea;
                             otherSideEnclosureNum = enclosureNum;
-                            thisEnclosure.spaceNames.push_back(state.dataHeatBal->space(state.dataSurface->Surface(surf.ExtBoundCond).spaceNum).Name);
-                            thisEnclosure.spaceNums.push_back(state.dataSurface->Surface(surf.ExtBoundCond).spaceNum);
-                            thisEnclosure.FloorArea += state.dataHeatBal->Zone(state.dataSurface->Surface(surf.ExtBoundCond).Zone).FloorArea;
+                            int otherSideSpaceNum = state.dataSurface->Surface(surf.ExtBoundCond).spaceNum;
+                            if (otherSideSpaceNum != surf.spaceNum) {
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:14815]: For #9233: Add other side space and floor area ONLY if the other side space is a different space.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 2]
```diff
@@ -14819,20 +14819,23 @@ namespace SurfaceGeometry {
...
                                 state.dataHeatBal->space(state.dataSurface->Surface(surf.ExtBoundCond).spaceNum).radiantEnclosureNum =
                                     otherSideEnclosureNum;
                             } else {
                                 thisEnclosure.ExtWindowArea += state.dataHeatBal->space(surf.spaceNum).extWindowArea;
                                 thisEnclosure.TotalSurfArea += state.dataHeatBal->space(surf.spaceNum).totalSurfArea;
-                                thisEnclosure.ExtWindowArea +=
-                                    state.dataHeatBal->space(state.dataSurface->Surface(surf.ExtBoundCond).spaceNum).extWindowArea;
-                                thisEnclosure.TotalSurfArea +=
-                                    state.dataHeatBal->space(state.dataSurface->Surface(surf.ExtBoundCond).spaceNum).totalSurfArea;
+                                if (otherSideSpaceNum != surf.spaceNum) {
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:14827]: For #9233: Same logic for additional solar enclosure values.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 3]
```diff
@@ -14892,11 +14895,11 @@ namespace SurfaceGeometry {
                                 std::string saveName = Enclosures(enclNum).Name;
                                 Enclosures(enclNum) = Enclosures(enclNum + 1);
                                 Enclosures(enclNum).Name = saveName;
-                                for (auto zNum : thisEnclosure.spaceNums) {
+                                for (int sNum : Enclosures(enclNum).spaceNums) {
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:14890]: For #9218: `thisEnclosure` was set above to `Enclosures(enclNum)` but then line 14896 shifts that, so it was the wrong enclosure to drive this for loop. This caused wrong enclosure numbers in some spaces which caused repeated merging of the same enclosure, eventually driving the number of enclosures to zero which caused the crash in the defect file.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 4]
```diff
@@ -14892,11 +14895,11 @@ namespace SurfaceGeometry {
...
                                     if (radiantSetup) {
-                                        state.dataHeatBal->space(zNum).radiantEnclosureNum = enclNum;
+                                        state.dataHeatBal->space(sNum).radiantEnclosureNum = enclNum;
```
Comment (mjwitte) [inline src/EnergyPlus/SurfaceGeometry.cc:14892]: Just some cleanup.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 5]
```diff
@@ -5990,6 +5990,856 @@ TEST_F(EnergyPlusFixture, HeatBalanceIntRadExchange_SetupEnclosuresWithAirBounda
     EXPECT_EQ(state->dataHeatBal->space(5).solarEnclosureNum, 1);
 }
 
+TEST_F(EnergyPlusFixture, HeatBalanceIntRadExchange_SetupEnclosuresWithAirBoundaries4)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:5993]: New unit test for #9218.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 6]
```diff
@@ -5990,6 +5990,856 @@ TEST_F(EnergyPlusFixture, HeatBalanceIntRadExchange_SetupEnclosuresWithAirBounda
...
+
+    // Check solar enclosures
+    for (int enclNum = 1; enclNum <= state->dataViewFactor->NumOfSolarEnclosures; ++enclNum) {
+        for (int spaceNum : state->dataViewFactor->EnclSolInfo(enclNum).spaceNums) {
+            EXPECT_EQ(state->dataHeatBal->space(spaceNum).solarEnclosureNum, enclNum);
+        }
+    }
+}
+
+TEST_F(EnergyPlusFixture, HeatBalanceIntRadExchange_SetupEnclosuresWithAirBoundaries5_SameSpace)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:6565]: New unit test for #9233.

**Review by mjwitte — COMMENTED** [review]
[CI results](https://github.com/NREL/EnergyPlus/pull/9237#commitcomment-63913169) on initial commit with new unit tests shows that the tests fail. With the fixes, the unit tests pass.

Code walkthrough.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:14815]: Makes perfect sense.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 2 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:14827]: Indeed.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:14890]: Wow.  Good catch!

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 4 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:14892]: OK

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc (hunk 5 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:5993]: Awesome!

**Review by Myoldmopar — APPROVED** [review]
This is fantastic!  Tightly contained changes, with a crystal clear code walkthrough, along with new unit tests.  :+1:

**Comment by Myoldmopar** [comment]
This looks perfect right now, but I will wait to hear that you are totally done with it before any local testing.  Thanks for the code walkthrough and documenting your process along the way in this PR conversation.

**Comment by mjwitte** [comment]
@Myoldmopar Yes, I'm finished with this PR. Merged in develop one last time.

**Comment by Myoldmopar** [comment]
Thanks @mjwitte , the primary CI machine went down for updates, but it's coming around now.  I will let a few more run but I assume this will merge soon.

**Comment by Myoldmopar** [comment]
Super happy here, merging.  Thanks @mjwitte

**Pull request merged**